src/tools/rebuild_mondb.cc

```diff
@@ -240,72 +240,69 @@ int update_osdmap(ObjectStore& fs, OSDSuperblock& sb, MonitorDBStore& ms)
   OSDMap osdmap;
   for (auto e = std::max(last_committed+1, sb.oldest_map);
        e <= sb.newest_map; e++) {
-    bool have_crc = false;
-    uint32_t crc = -1;
-    uint64_t features = 0;
-    // add inc maps
+    // add full maps
     {
-      const auto oid = OSD::get_inc_osdmap_pobject_name(e);
+      const auto oid = OSD::get_osdmap_pobject_name(e);
       bufferlist bl;
       int nread = fs.read(ch, oid, 0, 0, bl);
       if (nread <= 0) {
         cerr << "missing " << oid << std::endl;
         return -EINVAL;
       }
+      t->put(prefix, ms.combine_strings("full", e), bl);
+
+      auto p = bl.cbegin();
+      osdmap.decode(p);
+    }
+    // add inc maps
+    auto add_inc_result = [&] {
+      const auto oid = OSD::get_inc_osdmap_pobject_name(e);
+      bufferlist bl;
+      int nread = fs.read(ch, oid, 0, 0, bl);
+      if (nread <= 0) {
+        cout << "missing " << oid << std::endl;
+        return -ENOENT;
+      }
       t->put(prefix, e, bl);
 
       OSDMap::Incremental inc;
       auto p = bl.cbegin();
       inc.decode(p);
-      features = inc.encode_features | CEPH_FEATURE_RESERVED;
+      uint64_t features = inc.encode_features | CEPH_FEATURE_RESERVED;
       if (osdmap.get_epoch() && e > 1) {
         if (osdmap.apply_incremental(inc)) {
```

Comment:
If both inc and full osdmap exists, and we decode full osdmap first, 
then apply_incremental will fail with assert(inc.epoch == epoch+1)?

Reply:
@hjwsm1989 yeah. thanks for pointing this out. 

fixed and repushed.